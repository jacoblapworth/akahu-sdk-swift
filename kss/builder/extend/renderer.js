const { regexDictionary, types } = require('../parser/constants');
const tinycolor = require('tinycolor2');

module.exports = function(handlebars) {
	/**
	 * Extracts the variable properties from the line
	 * returns:
	 *  token name, token value, and optional [ token example ]
	 */
	handlebars.registerHelper('renderTokens', renderTokens);

	/**
	 * Compares the name to see if the container needs to be rendered as inverted.
	 * Similar to the isEqual
	 *
	 * e.g.
	 * {{#isInverted name}}
	 * 		ADD THIS
	 * {{else}} [optional]
	 * 		WILL EXECUTE THIS
	 * {{/isInverted}}
	 */
	handlebars.registerHelper('isInverted', isInverted);
};

const testString = 'To understand what recursion is, you must first understand recursion.';


function renderTokens(variables, block) {
	var newBlock = '';
	variables.split('\n').forEach(line => {
		var test;
		const matches = line.match(/[^:]+/g).map(match => match.trim());

		if (matches[1]) {
			this.token = {};
			this.token.name = matches[0];
			this.token.value = matches[1];
			this.token.title = matches[3];
			this.token.description = matches[4];
			if (matches[2] != undefined) {
				this.token.example = getExample(matches[2], this.token);
			}
			newBlock += block.fn(this);
		}
	});
	return newBlock;
}

function getExample(type, token) {
	switch(type) {
		case types.color:
			const codeTextClass = tinycolor(token.value).isDark() ?
				'ds-colour-code-light'
				: null;
			return `
					<div class="ds-colour-example" style="background-color:${token.value}">
						<code class="${codeTextClass}">${token.value}</code>
						<code class="${codeTextClass}">${token.name}</code>
					</div>`;
		case types.border:
			return `
				<div class="ds-colour-example" style="border:${token.value}">
					<code>${token.name}</code>
				</div>`;
		case types.shadow:
			return `
					<div class="ds-colour-example" style="box-shadow:${token.value}">
						<code>${token.name}</code>
					</div>`;
		case types.fontSize:
			return  `<span class="xui-text-truncated" style="font-size:${token.value};">Aa</span>`;
		case types.fontWeight:
			return `<span style="font-weight:${token.value};">Aa</span>`;
		case types.lineHeight:
			return `<span style="line-height:${token.value}">${testString}</span>`;
		case types.spacing:
			return `<div class="ds-spacing-example" style="height:${token.value};width:${token.value};"></div>`;
		default:
			return '';
	}
}

function isInverted(name, options) {
	return regexDictionary.inverted.test(name) ? options.fn(this) : options.inverse(this);
}
