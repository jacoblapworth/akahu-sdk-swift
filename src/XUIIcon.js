import React, {PropTypes} from 'react';
import cn from 'classnames';
import XUIIconBlob from './XUIIconBlob';
import Classes from 'xui-css-classes';
import icons from './iconData';

const iconNames = Object.keys(icons).reduce((object, icon, i) => {
		object[icon.replace('-','_').toUpperCase()] = icon;
		return object;
	}, {});
export const XUIIcons = iconNames;
const iconBlobId = 'xui-icon-blob-auto';
const iconSizes = {
	standard: '',
	large: Classes.Icon.LARGE,
	xlarge: Classes.Icon.XLARGE
};
const rotations = [
	0,
	90,
	180,
	270
];
const colors = Object.keys(Classes.Icon.Color).map(key=>{return key.toLowerCase()});
const propTypes = {
	/** @property {string} Defines the icon to be rendered */
	icon: PropTypes.oneOf(Object.keys(icons)).isRequired,
	/** @property {string} [className=''] Additional classes to be applied to the icon */
	className: PropTypes.string,
	/** @property {string} [size=''] Adds a size modifier to the icon */
	size: PropTypes.oneOf(Object.keys(iconSizes)),
	/** @property {string} [title=''] Title to be read by screen readers */
	title: PropTypes.string,
	/** @property {string} [desc=''] Description to be read by screen readers */
	desc: PropTypes.string,
	/** @property {string} [desc=''] Role to be applied to the SVG for screen readers */
	role: PropTypes.string,
	/** @property {string} [rotation=0] Adds a rotation modifier to the icon */
	rotation: PropTypes.oneOf(rotations),
	/** @property {string} [color] Adds a color modifier to the icon */
	color: PropTypes.oneOf(colors),
	/** @property {bool} [inline] Whether the inline class modifier should be added */
	inline: PropTypes.bool
};

const defaultProps = {
	size: 'standard',
	role: 'presentation'
};

export default function XUIIcon(props) {
	const { icon, className, size, title, desc, role, rotation, color, inline } = props;
	const classes = cn(
		Classes.Icon.BASE,
		className,
		iconSizes[size],
		Classes.Icon.Color[color ? color.toUpperCase() : null],
		{
			[Classes.Icon.ROTATE[rotation]] : rotation,
			[Classes.Icon.INLINE] : inline
		}
		);

	if (typeof(document) !== 'undefined' && !document.getElementById(iconBlobId)) {
		renderBlob();
	}

	const optionalTitle = title? <title>{ title }</title> : null;
	const optionalDescription = desc? <desc>{ desc }</desc> : null;

	return(
		<svg focusable="false" className={ classes }>
			{ optionalTitle }
			{ optionalDescription }
			<use xlinkHref={'#xui-icon-' + icon } role={ role }/>
		</svg>
	);
}

function renderBlob() {
	const blobEl = document.createElement('div');
	blobEl.id = iconBlobId;
	blobEl.innerHTML = XUIIconBlob;
	document.body.appendChild(blobEl);
}

XUIIcon.propTypes = propTypes;
XUIIcon.defaultProps = defaultProps;
