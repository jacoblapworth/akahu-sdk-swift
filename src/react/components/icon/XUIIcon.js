import PropTypes from 'prop-types';
import React from 'react';
import cn from 'classnames';
import '../helpers/xuiGlobalChecks';
import {
  baseClass,
  wrapperClass,
  wrapperSizeClasses,
  rotationClasses,
  colorClasses,
  iconSizeMultipliers,
} from './private/constants';

const XUIIcon = React.forwardRef((props, ref) => {
  const { className, qaHook, size, title, desc, role, rotation, color, icon, isBoxed } = props;

  const svgClasses = cn(
    baseClass,
    colorClasses[color],
    rotationClasses[rotation],
    !isBoxed && className, // If no wrapper is needed, apply custom classes directly on the SVG
  );

  const optionalTitle = title ? <title>{title}</title> : null;
  const optionalDescription = desc ? <desc>{desc}</desc> : null;
  const sizeMultiplier = iconSizeMultipliers[size] || 1;

  const svgElement = (
    <svg
      className={svgClasses}
      data-automationid={qaHook}
      focusable="false"
      height={icon.height * sizeMultiplier}
      ref={ref}
      viewBox={`0 0 ${icon.width} ${icon.height}`}
      width={icon.width * sizeMultiplier}
    >
      {optionalTitle}
      {optionalDescription}
      <path d={icon.path} role={role} />
    </svg>
  );

  return !isBoxed ? (
    svgElement
  ) : (
    <div
      className={cn(className, wrapperClass, wrapperSizeClasses[size])}
      data-automationid={qaHook && `${qaHook}-wrapper`}
    >
      {svgElement}
    </div>
  );
});

XUIIcon.propTypes = {
  className: PropTypes.string,
  /** Adds a color modifier to the icon */
  color: PropTypes.oneOf(Object.keys(colorClasses)),
  /** Description to be read by screen readers */
  desc: PropTypes.string,
  /** An object describing the path, width and height. */
  icon: PropTypes.shape({
    height: PropTypes.number.isRequired,
    path: PropTypes.string.isRequired,
    width: PropTypes.number.isRequired,
  }).isRequired,
  /** Whether the icon should be wrapped in a wrapper with a set size */
  isBoxed: PropTypes.bool,
  qaHook: PropTypes.string,
  /** Role to be applied to the SVG for screen readers */
  role: PropTypes.string,
  /** Adds a rotation modifier to the icon. Accepted values are 0 (default), 90, 180, 270 */
  rotation: PropTypes.oneOf([
    ...Object.keys(rotationClasses),
    ...Object.keys(rotationClasses).map(n => parseInt(n)),
  ]),
  /** Adds a size modifier to the icon */
  size: PropTypes.oneOf(Object.keys(wrapperSizeClasses)),
  /** Title to be read by screen readers */
  title: PropTypes.string,
};

XUIIcon.defaultProps = {
  role: 'presentation',
  size: 'medium',
};

export default XUIIcon;
