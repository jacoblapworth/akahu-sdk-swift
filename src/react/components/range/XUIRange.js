import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import XUIControlWrapper, { getAriaAttributes } from '../controlwrapper/XUIControlWrapper';
import { ns } from '../helpers/xuiClassNamespace';
import generateIds from '../controlwrapper/helpers';

import '../../../sass/7-components/_forms.range.scss';

const baseClass = `${ns}-rangeslider`;


export default class XUIRange extends PureComponent {
	wrapperIds = generateIds();
	render() {
		const {
			hintMessage,
			qaHook,
			onClick,
			onInput,
			leftElement,
			rightElement,
			name,
			isLabelHidden,
			label,
			id,
			step,
			size,
			defaultValue,
			min,
			max,
			isDisabled,
			validationMessage,
			isInvalid,
			containerClasses,
			inputClasses,
		} = this.props;

		return (
			<XUIControlWrapper
				wrapperIds={this.wrapperIds}
				isLabelHidden={isLabelHidden}
				label={label}
				validationMessage={validationMessage}
				hintMessage={hintMessage}
				qaHook={qaHook}
				isInvalid={isInvalid}
			>
				<div className={cn(`${baseClass}-container`, containerClasses)} >
					{leftElement}
					<input
						name={name}
						onClick={onClick}
						onInput={onInput}
						id={id}
						className={cn(
							`${baseClass}`,
							size && `${baseClass}-thumb-${size}`,
							isInvalid && `${baseClass}-is-invalid`,
							inputClasses,
						)}
						step={step}
						type="range"
						defaultValue={defaultValue}
						min={min}
						max={max}
						disabled={isDisabled}
						{...getAriaAttributes(this.wrapperIds, this.props)}
					/>
					{rightElement}
				</div>
			</XUIControlWrapper>
		);
	}
}

XUIRange.propTypes = {
	qaHook: PropTypes.string,
	/** Element on the left of the XUIRange component */
	leftElement: PropTypes.node,
	/** Element on the right of the XUIRange component */
	rightElement: PropTypes.node,
	/** Boolean to hide/show Input label */
	isLabelHidden: PropTypes.node,
	/** Input label */
	label: PropTypes.node,
	/** Name to be consumed by form objects etc */
	name: PropTypes.string,
	/** Hint string to be passed into the XUIControlWrapper */
	hintMessage: PropTypes.string,
	/** Id of Range Component */
	id: PropTypes.string,
	/** Sizing of the circle touch object */
	size: PropTypes.oneOf(['default', 'small', 'xsmall']),
	/** The incremement of the XUIRange slider value */
	step: PropTypes.number,
	/** Default value of the XUIRange component */
	defaultValue: PropTypes.string,
	/** Minimum value of the XUIRange component */
	min: PropTypes.number,
	/** Maximum value of the XUIRange component */
	max: PropTypes.number,
	/** Define a function for the onClick event of the slider being clicked */
	onClick: PropTypes.func,
	/** Define a function for the onInput event of when the value of the slider is set/changed */
	onInput: PropTypes.func,
	/** Disables the XUIRange component */
	isDisabled: PropTypes.bool,
	/** The message to show in validation */
	validationMessage: PropTypes.string,
	/** Displays that the XUIRange component is invalid */
	isInvalid: PropTypes.bool,
	/** Additional classes to be applied to the container */
	containerClasses: PropTypes.string,
	/** Additional classes to be applied to the input */
	inputClasses: PropTypes.string,
};
