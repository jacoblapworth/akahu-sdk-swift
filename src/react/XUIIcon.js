import XUIIcon, { XUIIcons } from './components/icon/XUIIcon';
import XUICustomIcon from './components/icon/XUICustomIcon';
import { ensureIconBlobOnPage } from './components/icon/XUIIconBlob';

export { XUIIcon as default, XUIIcons, XUICustomIcon, ensureIconBlobOnPage };
