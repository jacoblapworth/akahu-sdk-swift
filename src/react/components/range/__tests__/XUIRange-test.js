import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import XUIRange from '../XUIRange';
import uuidv4 from 'uuid/v4';
import XUIIcon from '../../../icon'

Enzyme.configure({ adapter: new Adapter() });

jest.mock('uuid/v4');
uuidv4.mockImplementation(() => 'testRangeId');

describe('Range', () => {

	// type range
	it('Simple case', () => {
		const component = shallow(<XUIRange />);
		const simpleRangeComponent = renderer.create(
			<XUIRange />
		);
		expect(simpleRangeComponent).toMatchSnapshot();
	});

	it('invalid', () => {
		const invalidRangeComponent = renderer.create(
			<XUIRange label='invalid'
				id="range1"
				isInvalid
				validationMessage="validationMessage"
				min={16}
				max={80} />
		);
		expect(invalidRangeComponent).toMatchSnapshot();
	});


	it('should be disabled', () => {
		const disabledRangeComponent = renderer.create(
			<XUIRange label='disabled'
				id="range1"
				validationMessage="validationMessage"
				min={16}
				max={80}
				isDisabled />
		);
		expect(disabledRangeComponent).toMatchSnapshot();
	});

	it('LeftElement and RightElement work', () => {
		const { default: plus } = require('@xero/xui-icon/icons/plus');
		const svgRangeComponent = renderer.create(
			<XUIRange label='svgs work'
				leftElement={<XUIIcon size="large" icon={plus} />}
				rightElement={<XUIIcon size="large" icon={plus} />}
				id="range1"
				hintMessage='hintMessage'
				validationMessage="validationMessage" />
		);
		expect(svgRangeComponent).toMatchSnapshot();
	});

	it('should call the onClick callback on the range component', () => {
		const onClick = jest.fn();
		const comp = mount(<XUIRange id="rangeComponent" onClick={onClick} />);

		comp.find('#rangeComponent').first().simulate('click');
		setTimeout( () => {
				expect(onSelect.mock.calls.length).toEqual(1);
		}, 0);
	});

	it('All options selected', () => {
		const { default: plus } = require('@xero/xui-icon/icons/plus');
		const allOptionsRangeComponent = renderer.create(
			<XUIRange label='All options selected'
				leftElement={<XUIIcon size='large' icon={plus} />}
				rightElement={<XUIIcon size='large' icon={plus} />}
				qaHook='qaHook'
				hintMessage='hintMessage'
				id="range1"
				inputClasses='inputClasses'
				containerClasses='containerClasses'
				isInvalid
				validationMessage='validationMessage'
				isDisabled
				min={16}
				max={80}
				size='xsmall'
				step={4} />
		);
		expect(allOptionsRangeComponent).toMatchSnapshot();
	});


});
