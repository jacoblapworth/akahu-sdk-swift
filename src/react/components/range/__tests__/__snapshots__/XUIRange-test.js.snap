// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range All options selected 1`] = `
<div
  className=""
>
  <label
    className="xui-text-label xui-fieldlabel-layout"
    data-automationid="qaHook--label"
    htmlFor="testRangeId-control"
  >
    All options selected
  </label>
  <div
    className="xui-rangeslider-container containerClasses"
  >
    <svg
      className="xui-icon"
      focusable="false"
      height={16}
      viewBox="0 0 12 12"
      width={16}
    >
      <path
        d="M7 5h5v2H7v5H5V7H0V5h5V0h2v5z"
        role="presentation"
      />
    </svg>
    <input
      aria-describedby="testRangeId-message"
      aria-invalid={true}
      className="xui-rangeslider xui-rangeslider-thumb-xsmall xui-rangeslider-is-invalid inputClasses"
      defaultValue="20"
      disabled={true}
      id="testRangeId-control"
      max="80"
      min="16"
      step="4"
      type="range"
    />
    <svg
      className="xui-icon"
      focusable="false"
      height={16}
      viewBox="0 0 12 12"
      width={16}
    >
      <path
        d="M7 5h5v2H7v5H5V7H0V5h5V0h2v5z"
        role="presentation"
      />
    </svg>
  </div>
  <div
    className="xui-validation xui-validation-layout xui-validation-is-invalid"
    data-automationid="qaHook--message"
    id="testRangeId-message"
    role="status"
  >
    <svg
      className="xui-icon"
      focusable="false"
      height={16}
      viewBox="0 0 16 16"
      width={16}
    >
      <path
        d="M8,16 C3.581722,16 0,12.418278 0,8 C0,3.581722 3.581722,0 8,0 C12.418278,0 16,3.581722 16,8 C16,12.418278 12.418278,16 8,16 Z M7,4 L7,9 L9,9 L9,4 L7,4 Z M7,10 L7,12 L9,12 L9,10 L7,10 Z"
        role="presentation"
      />
    </svg>
    validationMessage
  </div>
</div>
`;

exports[`Range LeftElement and RightElement work 1`] = `
<div
  className=""
>
  <label
    className="xui-text-label xui-fieldlabel-layout"
    htmlFor="testRangeId-control"
  >
    svgs work
  </label>
  <div
    className="xui-rangeslider-container"
  >
    <svg
      className="xui-icon"
      focusable="false"
      height={16}
      viewBox="0 0 12 12"
      width={16}
    >
      <path
        d="M7 5h5v2H7v5H5V7H0V5h5V0h2v5z"
        role="presentation"
      />
    </svg>
    <input
      aria-describedby="testRangeId-message"
      className="xui-rangeslider xui-rangeslider-thumb-medium"
      id="testRangeId-control"
      type="range"
    />
    <svg
      className="xui-icon"
      focusable="false"
      height={16}
      viewBox="0 0 12 12"
      width={16}
    >
      <path
        d="M7 5h5v2H7v5H5V7H0V5h5V0h2v5z"
        role="presentation"
      />
    </svg>
  </div>
  <div
    className="xui-validation xui-validation-layout"
    id="testRangeId-message"
    role="status"
  >
    hintMessage
  </div>
</div>
`;

exports[`Range Simple case 1`] = `
<div
  className=""
>
  <div
    className="xui-rangeslider-container"
  >
    <input
      className="xui-rangeslider xui-rangeslider-thumb-medium"
      id="testRangeId-control"
      type="range"
    />
  </div>
</div>
`;

exports[`Range invalid 1`] = `
<div
  className=""
>
  <label
    className="xui-text-label xui-fieldlabel-layout"
    htmlFor="testRangeId-control"
  >
    invalid
  </label>
  <div
    className="xui-rangeslider-container"
  >
    <input
      aria-describedby="testRangeId-message"
      aria-invalid={true}
      className="xui-rangeslider xui-rangeslider-thumb-medium xui-rangeslider-is-invalid"
      id="testRangeId-control"
      max={80}
      min={16}
      type="range"
    />
  </div>
  <div
    className="xui-validation xui-validation-layout xui-validation-is-invalid"
    id="testRangeId-message"
    role="status"
  >
    <svg
      className="xui-icon"
      focusable="false"
      height={16}
      viewBox="0 0 16 16"
      width={16}
    >
      <path
        d="M8,16 C3.581722,16 0,12.418278 0,8 C0,3.581722 3.581722,0 8,0 C12.418278,0 16,3.581722 16,8 C16,12.418278 12.418278,16 8,16 Z M7,4 L7,9 L9,9 L9,4 L7,4 Z M7,10 L7,12 L9,12 L9,10 L7,10 Z"
        role="presentation"
      />
    </svg>
    validationMessage
  </div>
</div>
`;

exports[`Range should be disabled 1`] = `
<div
  className=""
>
  <label
    className="xui-text-label xui-fieldlabel-layout"
    htmlFor="testRangeId-control"
  >
    disabled
  </label>
  <div
    className="xui-rangeslider-container"
  >
    <input
      className="xui-rangeslider xui-rangeslider-thumb-medium"
      disabled={true}
      id="testRangeId-control"
      max={80}
      min={16}
      type="range"
    />
  </div>
  <div
    className="xui-validation xui-validation-layout"
    id="testRangeId-message"
    role="status"
  />
</div>
`;
